[libc++][Android] Restrict use of mblen/towctrans/wctrans

rprichard opened the pull request:
These functions weren't added until API 26 (Android 8.0), but libc++ is supported for API 21 and up.

These APIs are undeclared as of r.android.com/3216959.

rprichard committed:
[libc++][Android] Restrict use of mblen/towctrans/wctrans

These functions weren't added until API 26 (Android 8.0), but libc++ is
supported for API 21 and up.

These APIs are undeclared as of r.android.com/3216959.

rprichard committed:
fix formatting

philnik777 reviewed (commented):
I'm fine with this, but I'd like @ldionne to approve as well.

rprichard committed:
Merge remote-tracking branch 'origin/main' into android-mblen-wctrans

ldionne reviewed (commented):
To understand, these Android API versions are like SDK versions, is that right? Does this basically mean that Android API 21 is the "SDK" released in 2014 as part of Android 5, so basically you'd support using a new libc++ against an Android SDK released in 2014?

What is the intended support policy for libc++ on Android? At what frequency is support for older platforms dropped? I'd just like to make sure that there is some reasonable policy in place to make sure we don't end up having to support ancient SDKs (which often doesn't add much value since mixing a new libc++ with an old SDK is a niche use case). For example, for Apple platforms we officially support only the latest released XCode (with the matching SDK).

rprichard commented:
+ @DanAlbert

> To understand, these Android API versions are like SDK versions, is that right? Does this basically mean that Android API 21 is the "SDK" released in 2014 as part of Android 5, so basically you'd support using a new libc++ against an Android SDK released in 2014?
> 
> What is the intended support policy for libc++ on Android? At what frequency is support for older platforms dropped? I'd just like to make sure that there is some reasonable policy in place to make sure we don't end up having to support ancient SDKs (which often doesn't add much value since mixing a new libc++ with an old SDK is a niche use case). For example, for Apple platforms we officially support only the latest released XCode (with the matching SDK).

It's more like the version of Android itself rather than the version of the SDK (NDK). Android 5.0 is Android Lollipop, which is also API level 21. (Android 5.1 is API level 22 instead.) (See https://apilevels.com.) Each Android release has three similar versions (API level, version number, and codename). The NDK contains a copy of libc++, and the NDK is on a separate release schedule from the Android OS. https://developer.android.com/ndk/downloads/revision_history

When the NDK cuts a new release, with a recent version of libc++, the new NDK supports targeting older Android versions. This lets app developers use the latest Clang and libc++ even though they still need to support old versions of Android.

I think the Android team would like libc++ to continue supporting an Android OS for as long as the NDK does. Currently the NDK supports API level 21 and up, but we expect to drop old API levels over time. (The minimum supported API level has been steadily marching up over the past several years. e.g. NDK r16 supported Android 4.0 and up -- API level 14.) I'll cc @DanAlbert who I think is more familiar with the policy on dropping support for old Android OSs.

rprichard commented:
@ldionne I think this PR is still waiting for review. It's needed before I can update the version of Clang used to test libc++ on Android.

ldionne reviewed (commented):
Sorry this fell between the cracks and I forgot to review again. Thanks for the clear explanation of Android NDK versions.

For these kinds of things in general, we tend to favour using `XFAIL: android-ndk-<version>` instead, since that prevents the accumulation of `#ifdef`s in the test suite. Would you be willing to use that instead?

rprichard committed:
Merge branch 'main' into android-mblen-wctrans

rprichard committed:
Replace #if with UNSUPPORTED.

We can't use XFAIL yet because the old sysroot used in the Android CI
Docker image still declares these functions even for pre-API 26. The
new sysroot we want to switch stop drops the declarations, unless the
code is targeting API 26 or newer.

rprichard commented:
> Sorry this fell between the cracks and I forgot to review again. Thanks for the clear explanation of Android NDK versions.
> 
> For these kinds of things in general, we tend to favour using `XFAIL: android-ndk-<version>` instead, since that prevents the accumulation of `#ifdef`s in the test suite. Would you be willing to use that instead?

We could use `UNSUPPORTED:` instead, which would work with both the old and the new sysroot. Is that OK?

Details:

We would need to do XFAIL based on both the target API and the NDK sysroot "version", but that's tricky because AFAIK the sysroot doesn't have a version on it.

The problem is that the NDK's sysroot happened to declare these functions even for old API levels, even though they weren't available in libc.so. It was an old hack to make libc++ happy, IIRC. The sysroot was modified to stop declaring the functions in https://r.android.com/3216959. Of course, they still _are_ declared for new enough API levels (26 and up).

i.e. This new XFAIL would work with the new sysroot:

```
// towctrans and wctrans were added in Android API 26.
// XFAIL: target={{.+}}-android{{(eabi)?(21|22|23|24|25)}}
```

... but it would break the libc++ Android CI until the sysroot in the Docker image was updated (because the tests are currently passing for API 21, which is tested in CI).

philnik777 commented:
> > Sorry this fell between the cracks and I forgot to review again. Thanks for the clear explanation of Android NDK versions.
> > For these kinds of things in general, we tend to favour using `XFAIL: android-ndk-<version>` instead, since that prevents the accumulation of `#ifdef`s in the test suite. Would you be willing to use that instead?
> 
> We could use `UNSUPPORTED:` instead, which would work with both the old and the new sysroot. Is that OK?

I think that should be fine with a `TODO` added so we don't forget to remove it again.

> Details:
> 
> We would need to do XFAIL based on both the target API and the NDK sysroot "version", but that's tricky because AFAIK the sysroot doesn't have a version on it.
> 
> The problem is that the NDK's sysroot happened to declare these functions even for old API levels, even though they weren't available in libc.so. It was an old hack to make libc++ happy, IIRC. The sysroot was modified to stop declaring the functions in https://r.android.com/3216959. Of course, they still _are_ declared for new enough API levels (26 and up).
> 
> i.e. This new XFAIL would work with the new sysroot:
> 
> ```
> // towctrans and wctrans were added in Android API 26.
> // XFAIL: target={{.+}}-android{{(eabi)?(21|22|23|24|25)}}
> ```
> 
> ... but it would break the libc++ Android CI until the sysroot in the Docker image was updated (because the tests are currently passing for API 21, which is tested in CI).

rprichard commented:
Ok, I've updated the tests to be `UNSUPPORTED` when the build target is Android API 25 or less. (libc++ requires Android API 21 or up, so the unsupported range is 21..25.)

philnik777 reviewed (approved):
If @ldionne doesn't have any further comments within the next few days I think this should be fine to land.
